Generate Software Bill of Materials from CI images

aanm opened the pull request:
Copy of https://github.com/cilium/cilium/pull/22191 to run the CI from a branch

aanm committed:
test GH workflow

Signed-off-by: André Martins <andre@cilium.io>

developer-guy commented on .github/workflows/build-images-ci.yaml:
```diff
@@ -244,6 +249,41 @@ jobs:
+      - name: Install Bom and generate SBOM
+        shell: bash
+        run: |
+          curl -sSL --remote-name-all https://github.com/kubernetes-sigs/bom/releases/download/v0.3.0/bom-linux-amd64.tar.gz
```
might be an option: https://github.com/puerco/bom-installer

developer-guy commented on .github/workflows/build-images-ci.yaml:
```diff
@@ -244,6 +249,41 @@ jobs:
+        env:
+          COSIGN_EXPERIMENTAL: "true"
+        run: |
+          cosign attest --type spdx -predicate sbom_source_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}
```
please use double quote syntax: _--predicate_

developer-guy commented on .github/workflows/build-images-ci.yaml:
```diff
@@ -244,6 +249,41 @@ jobs:
+          cosign attest --type spdx -predicate sbom_image_ci_race_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-race
+          cosign attest --type spdx -predicate sbom_image_ci_unstripped_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-unstripped
+
+      - name: Verify SBOM attestation
```
I'm not sure that you need to do verification here because you are the provider of the software and your responsibility is to sign and distribute it, so later on, the customers of your software should verify your software before deciding whether they use it (optionally).

sandipanpanda committed:
build: Generate SBOM during image release

Add a Software Bill of Materials (SBOM) generated from source code
and the image using the Kubernetes bom tool to the Cilium image
during image build.

Generating SBOM describing the source in the Cilium repository
using bom takes, on average, 10 minutes. As a result, the CI
build time increases by 30 minutes if we generate an SBOM
describing the source for all three CI images in Image CI Build
and the CI ultimately fails, throwing an error that no space
is left on the runner. So, SBOM generation from source was
removed from Image CI Build.

In theory, if you run "bom generate" in the same environment where
you are building (especially after building), all modules should be
there already downloaded, and bom can reuse them. But this does not
happen. One thing that bom will not do is download stuff into your
go directory. If a module is missing, bom will download it to
/tmp/spdx/gomod-scanner/, look at it there, and remove it. Even after
performing a "go mod download" before running "bom generate", the
downloaded modules are not being picked up by the go interpreter
when bom runs.

Fixes: #20712

Signed-off-by: Sandipan Panda <samparksandipan@gmail.com>
